radixdlt-core/radixdlt/src/main/java/com/radixdlt/tree/PMTBranch.java

```diff
@@ -122,13 +147,116 @@ public void insertNode(
       handleNewRemainder(key, val, acc, represent, read);
     } else {
       throw new IllegalStateException(
-          String.format("Unexpected subtree: %s", commonPath.whichRemainderIsLeft()));
+          String.format(UNEXPECTED_SUBTREE_ERROR_MSG, commonPath.whichRemainderIsLeft()));
+    }
+  }
+
+  @Override
+  public void removeNode(
+      PMTKey key, PMTAcc acc, Function<PMTNode, byte[]> represent, Function<byte[], PMTNode> read) {
+    final PMTPath commonPath = PMTPath.findCommonPath(this.getKey(), key);
+    if (commonPath.whichRemainderIsLeft() == PMTPath.RemainingSubtree.NONE) {
+      acc.remove(this);
+      var newBranch = new PMTBranch(this);
+      newBranch.setValue(null);
+      acc.add(newBranch);
+      acc.setTip(newBranch);
+    } else if (commonPath.whichRemainderIsLeft() == PMTPath.RemainingSubtree.NEW) {
+      var nextHash = this.getNextHash(key);
+      if (nextHash == null) {
+        acc.setNotFound();
+        return;
+      }
+      PMTNode currentChild = read.apply(nextHash);
+      currentChild.removeNode(key.getTailNibbles(), acc, represent, read);
+      PMTNode newChild = acc.getTip();
+      if (acc.notFound()) {
+        acc.setTip(null);
+      } else if (newChild == null) { // remove child
+        removeBranchChild(key, acc, read);
+      } else { // update child
+        updateBranchChild(key, acc, represent, newChild);
+      }
+    } else {
+      throw new IllegalStateException(
+          String.format(UNEXPECTED_SUBTREE_ERROR_MSG, commonPath.whichRemainderIsLeft()));
+    }
+  }
+
+  private void removeBranchChild(PMTKey key, PMTAcc acc, Function<byte[], PMTNode> read) {
+    if (numberOfChildren() == 1L) {
+      handleBranchWithOnlyOneChild(acc);
+    } else if (numberOfChildren() == 2L && (branchDoesNotHaveValue())) {
+      handleBranchWithTwoChildren(acc, read, key);
+    } else {
+      acc.remove(this);
+      var branchWithNext = new PMTBranch(this);
+      branchWithNext.setEmptyChild(key.getFirstNibble());
+      acc.add(branchWithNext);
+      acc.setTip(branchWithNext);
+    }
+  }
+
+  private void handleBranchWithOnlyOneChild(PMTAcc acc) {
+    if (branchDoesNotHaveValue()) {
+      acc.remove(this);
+      acc.setTip(null);
+    } else {
+      acc.remove(this);
+      var newLeaf = new PMTLeaf(new PMTKey(new byte[0]), this.getValue());
+      acc.add(newLeaf);
+      acc.setTip(newLeaf);
+    }
+  }
+
+  private void handleBranchWithTwoChildren(PMTAcc acc, Function<byte[], PMTNode> read, PMTKey key) {
```

Comment:
this is only for no value branch. Maybe it should be somehow merged with branchDoesNotHaveValue check?

Reply:
`branchDoesNotHaveValue` is used in another case as well.

Reply:
well, in one case it's outside, in the other inside.

Consider combining value + children computation for simplicity, as mentioned in comment above. These are different, but identify need for collapsing the branch well

Reply:
I'll rename `handleBranchWithTwoChildren` to `handleBranchWithTwoChildrenAndNoValue`